peer/upstream/upstream.go

```diff
@@ -0,0 +1,350 @@
+// Copyright (c) 2014-2019 Bitmark Inc.
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package upstream
+
+import (
+	"encoding/binary"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bitmark-inc/bitmarkd/blockdigest"
+	"github.com/bitmark-inc/bitmarkd/fault"
+	"github.com/bitmark-inc/bitmarkd/mode"
+	"github.com/bitmark-inc/bitmarkd/util"
+	"github.com/bitmark-inc/bitmarkd/zmqutil"
+	"github.com/bitmark-inc/logger"
+	zmq "github.com/pebbe/zmq4"
+)
+
+type UpstreamIntf interface {
```

Comment:
Why do you need to declare an interface for just one implementation?

Reply:
better for test

Reply:
Could you explain more? Better for test meaning?

Reply:
make a mock upstream to test higher levels without having to simulate a network

Reply:
We cannot mock without declaring an interface? I mean that can we mock a struct or something? I am not familiar with golang mock lib.

Reply:
I am not sure, can you help to find a solution?